lib/libutils/ext/include/confine_array_index.h

```diff
@@ -142,4 +142,19 @@ static inline size_t confine_array_index(size_t index, size_t size) {
   return safe_index;
 }
 #endif
+
+#ifdef __riscv
+static inline size_t confine_array_index(size_t index, size_t size) {
+	size_t safe_index = 0;
+	__asm__(
+	"slt	t0, %2, %1\n"
```

Comment:
Is this implementation speculation safe? I would have expected a barrier or something. Would you mind adding a short comment to give a hint as to what's going on?

By the way, you should probably not use explicit registers like `t0`, instead, you should use some `%x` allocated by the compiler.

Reply:
This may be useful: https://arxiv.org/abs/2206.04507

Reply:
This is a tricky subject, keep in mind that the `confine_array_index()` function is only supposed to be called with an index out of bound while speculating. So for a CPU that doesn't speculate the function will only see valid indexes.

Reply:
Yes, my logic is the same, return 0 for out-of-range indexes, otherwise return index itself, and the computation itself is not branching。

Reply:
Is this submission OK？

Reply:
I'm not sure this is a speculation safe implementation.

Reply:
> I'm not sure this is a speculation safe implementation.

How about we leave this function unimplemented until some real hardware is supported by OP-TEE? Then we'll have more information on whether or not the CPU is vulnerable and what mitigations are effective.

Reply:
It's needed by ldelf. I suppose we could have a naive C implementation as fallback until it's properly implemented on riscv.

Reply:
@jenswi-linaro yes, it is what I meant ("unimplemented" was not the proper word).

Reply:
@jenswi-linaro  hi，I don't quite understand “ a naive C implementation”。Does it mean to determine whether index is smaller than size use C language？

Reply:
Yes, to make things work, but without protection against side-channel attacks.

Reply:
Ok，I see。